refactor(common/lmlayer): starts a formal 'headless' mode

jahorton opened the pull request:
Depends on #2986.

Many thanks to @eddieantonio for his work on the original headless unit tests; they strongly illuminated my path in developing this PR.

This PR allows the lm-layer to be compiled for use on Node, rather than use in the DOM.  To do so, we now fill out the abstraction for initializing the "Worker" in a node-dependent manner and provide the necessary structures to emulate a `Worker` without actually making one.

See #2988 for unit tests against this new headless mode and #2992 for how it gets linked with KeymanWeb.

-----

It may be worth considering use of Node's `Worker` threads instead, but that adds more complexity (and node-reliance), which is pretty tricky to utilize well given that we currently can't use node-module type information within lm-layer code.

jahorton commented on common/predictive-text/virtualizedWorker.ts:
```diff
@@ -0,0 +1,86 @@
+       * Among other things, this will allow the worker to use its internal namespaces without issue.
+       */
+      let msgCommand = "onmessage(" + msgJSON + ")";
+      this._workerContext.__importScriptString(msgCommand);
```
The one concern I have here:  when the virtualized `Worker` calls its `postMessage`, that will call `this.workerPostMessage`, escaping the virtualized scope... but chances are that the worker's context will still be the `global` object at that time.

As long as we don't need to access that in callbacks, we should be fine.  In fact, it _is_ fine for now, with the LMLayer's current state.  Just noting a potential limitation of this approach.  (It's still better than losing access to the `Worker`'s `self` instance!  `Worker`s can usually afford to be far messier since they're self-contained.)

I did consider using a similar approach to reset to the original `global`, but there's the little detail that we'd need to call a method on the property of an instance - which would force us to (at least, temporarily) manipulate the global scope so that it'd be accessible.  In the worst case, that's a bridge we can cross if and when we come to it.

darcywong00 commented on common/predictive-text/build.sh:
```diff
@@ -45,13 +45,32 @@ build ( ) {
 
+# Builds the top-level JavaScript file (the second stage of compilation)
+build-main-headless () {
+  npm run tsc -- -p ./tsconfig.json || fail "Could not build top-level JavaScript file."
```
Do you want a different fail message from l.64?

eddieantonio commented on common/predictive-text/build.sh:
```diff
@@ -45,13 +45,32 @@ build ( ) {
 }
 
+# Build the worker and the main script for headless use.
+build_headless ( ) {
```
Style nitpick:

```suggestion
build-headless () {
```

(This is consistent with the rest of the build script.)

eddieantonio commented on common/predictive-text/build.sh:
```diff
@@ -180,7 +199,8 @@ if (( fetch_deps )); then
 fi
 
-build || fail "Compilation failed."
+build || fail "Browser-oriented compilation failed."
+build_headless || fail "Headless compilation failed."
```
If following my previous suggestion, then this need to be changed too:

```suggestion
build-browser || fail "Browser-oriented compilation failed."
build-headless || fail "Headless compilation failed."
```

eddieantonio commented on common/predictive-text/build.sh:
```diff
@@ -180,7 +199,8 @@ if (( fetch_deps )); then
   npm install --no-optional
 fi
 
```
We may want an explicit common build step called something like `build-wrapped-worker` code. The results of this is later consumed by `build-browser` and `build-headless`.

```suggestion

build-wrapped-worker || fail "LMLayer worker compilation failed"
```

jahorton replied:
I haven't gotten to this point yet, but I think I've covered the others now.

I'll work on that one in the morning, I suppose.  Shouldn't take long when I'm fresh.

eddieantonio commented on common/predictive-text/tsconfig.json:
```diff
@@ -1,17 +1,18 @@
     "allowJs": true,
+    "outFile": "build/headless.js",
     "module": "none",
-    "outFile": "./build/index.js",
     "inlineSources": true,
     "sourceMap": true,
     "target": "es5",
-    "lib": ["es6", "dom"]
+    "lib": ["es6"]
   },
   "include" : [
     "./*.ts"
   ],
   "exclude" : [
+    "web-defaultWorker.ts",
```
👍

eddieantonio commented on common/predictive-text/virtualizedWorker.ts:
```diff
@@ -0,0 +1,86 @@
+
+namespace com.keyman.text.prediction {
+  // Unfortunately, importing like this loses all type information for the modules.
+  // It's that, or massively overhaul lm-layer to be module-based instead of
+  // namespace-based.
```
```typescript
// TODO
```

:joy: :joy: :joy:

eddieantonio commented on common/predictive-text/virtualizedWorker.ts:
```diff
@@ -0,0 +1,86 @@
+
+    postMessage: (message: any, options: PostMessageOptions) => void;
+
+    importScripts() {
+      /* Use of vm.createContext and script.runInContext allow us to avoid
+        * polluting the global scope with imports.  When we throw away the
+        * context object, imported scripts will be automatically GC'd.
+        */
+      for(var i=0; i < arguments.length; i++) {
+        this.__importScriptString(fs.readFileSync(arguments[i]) as any as string);
+      }
+    }
```
Doesn't `fs.readFileSync()` without an `encoding` parameter [return `Buffer` by default[(https://nodejs.org/api/fs.html#fs_fs_readfilesync_path_options)? Either way, I'm suggesting these changes to better take advantage of TypeScript features, and have less uses of `as any as string` typing hacks.

```suggestion
    importScripts(...scriptNames: string[]) {
      /* Use of vm.createContext and script.runInContext allow us to avoid
       * polluting the global scope with imports.  When we throw away the
       * context object, imported scripts will be automatically GC'd.
       */
      for (let script of scriptNames) {
        this.__importScriptString(fs.readFileSync(script, "UTF-8"));
      }
    }
```

eddieantonio commented on common/predictive-text/virtualizedWorker.ts:
```diff
@@ -0,0 +1,86 @@
+    }
+
+    // Sends the worker's postMessage messages to the appropriate `onmessage` handler.
+    private workerPostMessage(message: any, options: any) {
```
Please use the `unknown` type instead:

```suggestion
    private workerPostMessage(message: unknown) {
```

[`unknown` is a type-safe version of `any`](https://mariusschulz.com/blog/the-unknown-type-in-typescript), useful when you don't have any clue what the message actually is.

eddieantonio commented on common/predictive-text/virtualizedWorker.ts:
```diff
@@ -0,0 +1,86 @@
+      this.self = this;
+    }
+
+    postMessage: (message: any, options: PostMessageOptions) => void;
```
Is `options` ever used? If not, I suggest removing it from the type signature for now:

```suggestion
    postMessage: (message: any) => void;
```

eddieantonio commented on common/predictive-text/virtualizedWorker.ts:
```diff
@@ -0,0 +1,86 @@
+      }
+    }
+
+    onmessage: (this: Worker, ev: MessageEvent) => any;
```
I forgot why this class does not implement `onmessage()`, but then I remembered how my own code works! :joy:

I suggest having a comment similar to this one to avoid future confusion:

```suggestion
    /* This is assigned by LMLayerWorker.install() during initialization */
    onmessage: (this: Worker, ev: MessageEvent) => any;
```

eddieantonio commented on common/predictive-text/worker-interface.d.ts:
```diff
@@ -18,7 +18,7 @@ interface PostMessageOptions {
 interface Worker {
   onmessage: ((this: Worker, ev: MessageEvent) => any) | null;
-  postMessage(message: any, transfer: Transferable[]): void;
+  postMessage(message: any, transfer: any[]): void;
   postMessage(message: any, options?: PostMessageOptions): void;
```
Once again, are these method overloads strictly necessary? What's the minimal interface required to be a valid `WebWorker` and provide the functionality required by the `LMLayer` worker?

Is this sufficient?

```suggestion
  postMessage(message: any): void;
```

jahorton commented:
Looks like I accidentally merged in the unit tests from #2988 into this PR while addressing PR comments.  You may want to check their code there, since they should still be viewable separately there.  Fortunately, this only added two unit test classes - the changes were completely isolated - so the extra content should be easy to filter out.
